dxr/plugins/python/analysis.py

```diff
@@ -202,4 +203,70 @@ def analyze_import(self, node):
                     if package_path:
                         import_name = package_path + '.' + import_name
 
-            self.tree_analysis.names[absolute_local_name] = import_name
+            import_tuple = Import(absolute_name=absolute_local_name,
+                                  line=node.lineno, col=node.col_offset,
+                                  import_name=import_name)
+            self.tree_analysis.definition_tree.add(import_tuple)
+
+
+# Namedtuples that are stored in DefinitionTree.
+Definition = namedtuple('Definition', ('absolute_name', 'line', 'col', 'path'))
+Import = namedtuple('Import', ('absolute_name', 'line', 'col', 'import_name'))
+
+
+class DefinitionTree(object):
+    """Stores definitions across the entire project, such as classes and
+    functions, as well as the imports that link them together.
+
+    """
+    def __init__(self):
+        self.modules = {}
+
+    def add(self, definition):
+        """Add a definition to the tree. """
+        names = definition.absolute_name.split('.')
+        scope = self.modules
+
+        for name in names[:-1]:
+            scope = scope.setdefault(name, {})
+        scope[names[-1]] = definition
+
+    def get(self, name):
+        """Fetch the definition for the given name, following imports if
+        necessary.
+
+        If the definition is not found, return None. Top-level modules
+        that have not been added to the tree are assumed external and
+        valid, and definitions will be returned for them with no path
+        or position data.
+        """
+        names = name.split('.')
+        module_name = names.pop(0)
+        value = self.modules.get(module_name, None)
+
+        # If we don't even find the top-level module, assume this is a
+        # built-in or external library.
+        if not value:
+            return Definition(name)
+
+        for name in names:
+            value = value.get(name, None)
+            if not value:
+                return None
+
+            while isinstance(value, Import):  # Follow imports.
+                value = self.get(value.import_name)
+
+        return value
+
+    def normalize_name(self, absolute_name):
+        """Given a name, figure out the actual module that the thing
+        that name points to lives and return that name.
+
+        For example, if you have `from os import path` in a module
```

Comment:
Ah, so this returns the canonical name. This example was super-spiffy helpful to me. More like this! :-)